File: music_car/android/app/src/main/java/com/icar/musiccar/engine/DanceEngine.kt
/**
 * 音乐小车 — 时间轴编舞引擎（Android / Kotlin）
 *
 * 每 50ms（20Hz）检查播放进度，在时间戳到达时通过 ApiClient 发送 HTTP 控制指令。
 *
 * 用法:
 *   val engine = DanceEngine(routine, ApiClient.getInstance())
 *   engine.onProgress = { elapsed, total -> ... }
 *   engine.onComplete = { ... }
 *   engine.start()
 */

package com.icar.musiccar.engine

import android.os.Handler
import android.os.Looper
import com.icar.musiccar.models.CmdRequest
import com.icar.musiccar.models.DanceMove
import com.icar.musiccar.models.DanceRoutine
import com.icar.musiccar.models.MoveTypes
import com.icar.musiccar.network.ApiClient
import kotlinx.coroutines.*

class DanceEngine(
    private val routine: DanceRoutine,
    private val api: ApiClient
) {
    private val handler = Handler(Looper.getMainLooper())
    private var moveIndex: Int = 0
    private var startTime: Long = 0
    private var running: Boolean = false
    private var currentMove: DanceMove? = null
    private var currentMoveEndTime: Long = 0
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private var tickJob: Job? = null

    /** 进度回调 (elapsed_ms, total_ms) */
    var onProgress: ((elapsed: Long, total: Long) -> Unit)? = null
    /** 完成回调 */
    var onComplete: (() -> Unit)? = null
    /** 错误回调 */
    var onError: ((msg: String) -> Unit)? = null
    /** 动作触发回调 */
    var onMove: ((move: DanceMove, index: Int) -> Unit)? = null

    /** 是否正在运行 */
    val isRunning: Boolean get() = running

    /** 当前进度百分比 */
    val progress: Int
        get() {
            if (!running) return 0
            val elapsed = System.currentTimeMillis() - startTime
            return minOf(100, (elapsed * 100 / routine.duration).toInt())
        }

    /** 开始执行编舞 */
    suspend fun start() {
        moveIndex = 0
        running = true
        currentMove = null
        startTime = System.currentTimeMillis()

        // 20Hz 主循环
        tickJob = scope.launch {
            while (isActive && running) {
                val elapsed = System.currentTimeMillis() - startTime

                // 进度回调
                onProgress?.invoke(elapsed, routine.duration)

                // 当前动作到期 → 自动停止
                currentMove?.let { move ->
                    if (elapsed >= currentMoveEndTime) {
                        sendCmd(CmdRequest(type = MoveTypes.STOP))
                        currentMove = null
                    }
                }

                // 触发新动作
                while (
                    moveIndex < routine.moves.size &&
                    elapsed >= routine.moves[moveIndex].timestamp
                ) {
                    val move = routine.moves[moveIndex]
                    executeMove(move, moveIndex)
                    moveIndex++
                }

                // 编舞结束
                if (moveIndex >= routine.moves.size && currentMove == null) {
                    stop()
                    onComplete?.invoke()
                    break
                }

                delay(50) // 20Hz
            }
        }
    }

    /** 停止执行 */
    fun stop() {
        running = false
        tickJob?.cancel()
        sendCmd(CmdRequest(type = MoveTypes.STOP))
    }

    // ── 内部 ──────────────────────────────────

    private fun executeMove(move: DanceMove, index: Int) {
        try {
            onMove?.invoke(move, index)

            when (move.type) {
                MoveTypes.MOVE -> {
                    sendCmd(CmdRequest(
                        type = MoveTypes.MOVE,
                        linear = move.linear,
                        angular = move.angular
                    ))
                    if (move.duration > 0) {
                        currentMove = move
                        currentMoveEndTime = startTime + move.timestamp + move.duration
                    }
                }
                MoveTypes.SPIN -> {
                    sendCmd(CmdRequest(
                        type = MoveTypes.MOVE,
                        linear = 0.0,
                        angular = move.angular
                    ))
                    if (move.duration > 0) {
                        currentMove = move
                        currentMoveEndTime = startTime + move.timestamp + move.duration
                    }
                }
                MoveTypes.STOP -> {
                    sendCmd(CmdRequest(type = MoveTypes.STOP))
                    currentMove = null
                }
                MoveTypes.PAUSE -> {
                    // pause 不做任何事，依赖上一个 duration 到期自动 stop
                    currentMove = null
                }
            }
        } catch (e: Exception) {
            onError?.invoke("动作 #$index 执行失败: ${e.message}")
        }
    }

    private fun sendCmd(cmd: CmdRequest) {
        scope.launch {
            try {
                api.postCmd(cmd)
            } catch (e: Exception) {
                onError?.invoke("指令发送失败: ${e.message}")
            }
        }
    }
}

File: music_car/ios/MusicCar/Engine/DanceEngine.swift
/**
 * 音乐小车 — 时间轴编舞引擎（iOS / Swift）
 *
 * 每 50ms（20Hz）检查播放进度，在时间戳到达时通过 ApiClient 发送 HTTP 控制指令。
 *
 * 用法:
 *   let engine = DanceEngine(routine: routine, api: ApiClient.shared)
 *   engine.onProgress = { elapsed, total in ... }
 *   engine.onComplete = { ... }
 *   await engine.start()
 */

import Foundation

@MainActor
class DanceEngine {
    private let routine: DanceRoutine
    private let api: ApiClient
    private var moveIndex: Int = 0
    private var startTime: TimeInterval = 0
    private var running: Bool = false
    private var currentMove: DanceMove? = nil
    private var currentMoveEndTime: TimeInterval = 0
    private var timer: Timer?
    
    /// 进度回调 (elapsed_ms, total_ms)
    var onProgress: ((Int64, Int64) -> Void)?
    /// 完成回调
    var onComplete: (() -> Void)?
    /// 错误回调
    var onError: ((String) -> Void)?
    /// 动作触发回调
    var onMove: ((DanceMove, Int) -> Void)?
    
    /// 是否正在运行
    var isRunning: Bool { running }
    
    /// 当前进度百分比
    var progress: Int {
        guard running else { return 0 }
        let elapsed = Date().timeIntervalSince1970 * 1000 - startTime
        return min(100, Int(elapsed * 100 / Double(routine.duration)))
    }
    
    init(routine: DanceRoutine, api: ApiClient = .shared) {
        self.routine = routine
        self.api = api
    }
    
    /// 开始执行编舞
    func start() async {
        moveIndex = 0
        running = true
        currentMove = nil
        startTime = Date().timeIntervalSince1970 * 1000
        
        // 20Hz 主循环
        timer = Timer.scheduledTimer(withTimeInterval: 1.0 / Double(DANCE_TICK_HZ), repeats: true) { [weak self] _ in
            guard let self = self, self.running else { return }
            
            Task { @MainActor in
                let elapsed = Int64(Date().timeIntervalSince1970 * 1000 - self.startTime)
                
                // 进度回调
                self.onProgress?(elapsed, self.routine.duration)
                
                // 当前动作到期 → 自动停止
                if let move = self.currentMove, Double(elapsed) >= self.currentMoveEndTime {
                    try? await self.api.postCmd(CmdRequest(type: MoveType.stop))
                    self.currentMove = nil
                }
                
                // 触发新动作
                while self.moveIndex < self.routine.moves.count &&
                      elapsed >= self.routine.moves[self.moveIndex].timestamp {
                    let move = self.routine.moves[self.moveIndex]
                    self.executeMove(move, index: self.moveIndex)
                    self.moveIndex += 1
                }
                
                // 编舞结束
                if self.moveIndex >= self.routine.moves.count && self.currentMove == nil {
                    self.stop()
                    self.onComplete?()
                }
            }
        }
    }
    
    /// 停止执行
    func stop() {
        running = false
        timer?.invalidate()
        timer = nil
        Task {
            try? await api.postCmd(CmdRequest(type: MoveType.stop))
        }
    }
    
    // ── 内部 ──────────────────────────────────
    
    private func executeMove(_ move: DanceMove, index: Int) {
        onMove?(move, index)
        
        Task {
            do {
                switch move.type {
                case MoveType.move:
                    try await api.postCmd(CmdRequest(
                        type: MoveType.move,
                        linear: move.linear,
                        angular: move.angular
                    ))
                    if move.duration > 0 {
                        currentMove = move
                        currentMoveEndTime = startTime + Double(move.timestamp) + Double(move.duration)
                    }
                    
                case MoveType.spin:
                    try await api.postCmd(CmdRequest(
                        type: MoveType.move,
                        linear: 0.0,
                        angular: move.angular
                    ))
                    if move.duration > 0 {
                        currentMove = move
                        currentMoveEndTime = startTime + Double(move.timestamp) + Double(move.duration)
                    }
                    
                case MoveType.stop:
                    try await api.postCmd(CmdRequest(type: MoveType.stop))
                    currentMove = nil
                    
                case MoveType.pause:
                    // pause 不做任何事
                    currentMove = nil
                    
                default:
                    break
                }
            } catch {
                onError?("动作 #\(index) 执行失败: \(error.localizedDescription)")
            }
        }
    }
}
